### PR TITLE
Solves an issue when IDP is planning optional shortest paths

### PR DESCRIPTION
**Pull request by systay**
The query 

`MATCH (a:A), (b:B) OPTIONAL MATCH p = shortestPath( (a)-[*]->(b) ) RETURN p`

was making IDP blow up

### CODE REVIEW

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/logical/QueryPlannerConfiguration.scala [hunk 1]
```diff
@@ -67,16 +67,17 @@ case class QueryPlannerConfiguration(leafPlanners: LeafPlannerList,
...
 }
 
-case class QueryPlannerKit(qg: QueryGraph,
-                           select: LogicalPlan => LogicalPlan,
+case class QueryPlannerKit(select: (LogicalPlan, QueryGraph) => LogicalPlan,
                            projectAllEndpoints: (LogicalPlan, QueryGraph) => LogicalPlan,
-                           pickBest: CandidateSelector)
+                           pickBest: CandidateSelector) {
+  def select(plans: Iterable[Seq[LogicalPlan]], qg: QueryGraph):Iterable[Seq[LogicalPlan]] =
+    plans.map(_.map( plan => select(plan, qg) ))
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/logical/QueryPlannerConfiguration.scala]: Code formatting is a bit off; padding spaces in function call and missing padding space around return type colon.

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/logical/idp/IDPQueryGraphSolver.scala [hunk 2]
```diff
@@ -56,39 +56,40 @@ case class IDPQueryGraphSolver(monitor: IDPQueryGraphSolverMonitor,
...
     }
 
-  private def planEmptyComponent(queryGraph: QueryGraph)(implicit context: LogicalPlanningContext, leafPlan: Option[LogicalPlan], kitFactory: (QueryGraph) => QueryPlannerKit): Seq[LogicalPlan] = {
+  private def planEmptyComponent(queryGraph: QueryGraph)(implicit context: LogicalPlanningContext, leafPlan: Option[LogicalPlan], kit: QueryPlannerKit): Seq[LogicalPlan] = {
     val plan = if (queryGraph.argumentIds.isEmpty)
       context.logicalPlanProducer.planSingleRow()
     else
       context.logicalPlanProducer.planQueryArgumentRow(queryGraph)
-    val result: LogicalPlan = kitFactory(queryGraph).select(plan)
+    val result: LogicalPlan = kit.select(plan, queryGraph)
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/logical/idp/IDPQueryGraphSolver.scala:91]: `result` shouldn't need to have a type annotation.

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/logical/idp/IDPQueryGraphSolver.scala [hunk 3]
```diff
@@ -98,10 +99,10 @@ case class IDPQueryGraphSolver(monitor: IDPQueryGraphSolverMonitor,
     // TODO: Investigate dropping leafPlanWeHopeToGetAwayWithIgnoring argument
     val leaves = leafPlanFinder(config, qg)
 
-    if (qg.patternRelationships.size > 0) {
+    if (qg.patternRelationships.nonEmpty) {
 
       val generators = solvers.map(_(qg))
-      val selectingGenerators = generators.map(_.map(kit.select))
+      val selectingGenerators = generators.map(_.map( plan => kit.select(plan, qg) ))
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/logical/idp/IDPQueryGraphSolver.scala]: Code format incorrect: padding space in function call.

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/logical/leafPlanOptions.scala [hunk 4]
```diff
@@ -26,11 +26,11 @@ import org.neo4j.cypher.internal.compiler.v2_3.planner.logical.plans.{LogicalLea
 object leafPlanOptions extends LogicalLeafPlan.Finder {
 
   def apply(config: QueryPlannerConfiguration, queryGraph: QueryGraph)(implicit context: LogicalPlanningContext): Set[LogicalPlan] = {
-    val kit = config.toKit(queryGraph)
+    val kit: QueryPlannerKit = config.toKit()
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/logical/leafPlanOptions.scala]: Unnecessary type annotation.

**Comment by pontusmelke** [comment]
@systay remove the WIP from commit message

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/logical/QueryPlannerConfiguration.scala [hunk 5]
```diff
@@ -67,16 +67,17 @@ case class QueryPlannerConfiguration(leafPlanners: LeafPlannerList,
                                      optionalSolvers: Seq[OptionalSolver],
                                      pickBestCandidate: LogicalPlanningFunction0[CandidateSelector]) {
 
-  def toKit(qg: QueryGraph)(implicit context: LogicalPlanningContext): QueryPlannerKit =
+  def toKit()(implicit context: LogicalPlanningContext): QueryPlannerKit =
     QueryPlannerKit(
-      qg = qg,
-      select = plan => applySelections(plan, qg),
+      select = (plan: LogicalPlan, qg: QueryGraph) => applySelections(plan, qg),
```
Comment (davidegrohmann) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/logical/QueryPlannerConfiguration.scala:72]: +1

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/logical/idp/IDPQueryGraphSolver.scala [hunk 6]
```diff
@@ -56,39 +56,39 @@ case class IDPQueryGraphSolver(monitor: IDPQueryGraphSolverMonitor,
...
-  private def selectShortestPath(kit: QueryPlannerKit, initialPlan: LogicalPlan)(implicit context: LogicalPlanningContext): LogicalPlan =
-    kit.qg.shortestPathPatterns.foldLeft(kit.select(initialPlan)) {
-      case (plan, sp) if sp.isFindableFrom(plan.availableSymbols) => kit.select(context.logicalPlanProducer.planShortestPaths(plan, sp))
+    kit.copy(select = selectShortestPath(kit, _, _))
+
+  private def selectShortestPath(kit: QueryPlannerKit, initialPlan: LogicalPlan, qg: QueryGraph)
+                                (implicit context: LogicalPlanningContext): LogicalPlan =
+    qg.shortestPathPatterns.foldLeft(kit.select(initialPlan, qg)) {
+      case (plan, sp) if sp.isFindableFrom(plan.availableSymbols) =>
+        val shortestPath: LogicalPlan = context.logicalPlanProducer.planShortestPaths(plan, sp)
```
Comment (davidegrohmann) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/planner/logical/idp/IDPQueryGraphSolver.scala]: type annotation can be removed here

**Pull request merged**